Expressions as annotation values (#586)
cap-js/docs: cds/cdl.md

  "@aDecimal": 11.1,
  "@aSymbol": {"#":"foo"},
  "@aReference": {"=":"foo.bar"},
  "@anArray": [ /* … */ ],
  "@anExpression": { /* see next section */ }
}
```

::: tip
In contrast to references in expressions (see next section), plain references aren't checked or resolved
by CDS parsers or linkers. They're interpreted and evaluated only on consumption-specific modules.
For example, for SAP Fiori models, it's the _4odata_ and _2edm(x)_ processors.
:::

### Expressions as Annotation Values (beta) {#expressions-as-annotation-values}

::: warning
Expressions in annotation values are released as beta feature.

> We provide an early preview for the functionality to allow you to experiment with it and provide feedback.

Reviewer: ```suggestion
We provide an early preview of this functionality to allow you to experiment with it and provide feedback.
```

```cds
@anExpression: ( foo.bar * 11 )
```

Syntactically, the same expressions are supported as in a select item or in the where clause of a query,
except subqueries. The expression can of course also be a single reference or a simple value:

```cds
@aRefExpr: ( foo.bar )
@aValueExpr: ( 11 )
```

Some advantage of using expressions as "first class" annotation values are:
* syntax and references are checked by the compiler
* code completion
* (planned for a later release) automatic path rewriting in propagated annotations (see below)
* (planned for a later release) automatic translation of expressions in OData annotations (see below)

#### Name resolution

Each path in the expression is checked. For an annotation assigned to an entity,
the first path step is resolved as element of the entity. For an annotation assigned
to an entity element, the first path step is resolved as the annotated element or its siblings.
If the annotation is assigned to a subelement of a structured element, the top level
elements of the entity can be accessed via `$self`.
A parameter `par` of a parametrized entity can be accessed like in a query with `:par`.

> If a path cannot be resolved successfully, compilation fails with an error.

Reviewer: I feel as if this paragraph should be a bullet list.

In contrast to `@aReference: foo.bar`, a single reference written as expression `@aRefExpr: ( foo.bar )`
is checked by the compiler.

#### CSN Representation

In CSN, the expression is represented as a record with two properties:
* A string representation of the expression is stored in property `=`.
* A tokenized representation of the expression is stored in one of the properties

> `xpr`, `ref`, `val`, `func`, or `list` (like if the expression was written in a query).

Reviewer: Do you want to list _all_? That would be:

```
  'ref', 'xpr', 'list', 'literal', 'val',
  '#', 'func', 'args', 'SELECT', 'SET',
  'cast',
```

Author: No, actually not.

```json
{
  "@anExpression": {
    "=": "foo.bar * 11",
    "xpr": [ {"ref": ["foo", "bar"]}, "*", {"value": 11} ]
  },
  "@aRefExpr": {
    "=": "foo.bar",
    "ref": ["foo", "bar"]
  },
  "@aValueExpr": {
    "=": "11",
    "val": 11
  }
}
```

Note the different CSN representations for a plain value `"@anInteger": 11` and a value written
as expression `@aValueExpr: ( 11 )`, respectively.

#### Propagation

[Annotations are propagated](#annotation-propagation) in views/projections, in includes, or along type references.
Currently, paths are not rewritten in propagated annotations. In a projection, for example,
all elements used in an annotation expression must be projected without renaming.
Thus, for the time being we recommend to use the feature mainly in top-level projections.

Example:

```cds
@Foo1 : (a)
@Foo2 : (a + b)
@Foo3 : (a + c)
entity E {
  a : Integer;
  b : Integer;
  c : Integer;
}

entity P as projection on E {
  a,

>   b as x

Reviewer: ```suggestion
  b as x,
```

Author: Don't like the look of that extra "," ...

* code completion
* (planned for a later release) automatic path rewriting in propagated annotations (see below)
* (planned for a later release) automatic translation of expressions in OData annotations (see below)

#### Name resolution

Each path in the expression is checked. For an annotation assigned to an entity,
the first path step is resolved as element of the entity. For an annotation assigned
to an entity element, the first path step is resolved as the annotated element or its siblings.
If the annotation is assigned to a subelement of a structured element, the top level
elements of the entity can be accessed via `$self`.
A parameter `par` of a parametrized entity can be accessed like in a query with `:par`.
If a path cannot be resolved successfully, compilation fails with an error.

In contrast to `@aReference: foo.bar`, a single reference written as expression `@aRefExpr: ( foo.bar )`
is checked by the compiler.

#### CSN Representation

In CSN, the expression is represented as a record with two properties:
* A string representation of the expression is stored in property `=`.
* A tokenized representation of the expression is stored in one of the properties

> `xpr`, `ref`, `val`, `func`, or `list` (like if the expression was written in a query).

Author: ```suggestion
`xpr`, `ref`, `val`, `func`, etc. (like if the expression was written in a query).
```

```json
{
  "@anExpression": {
    "=": "foo.bar * 11",
    "xpr": [ {"ref": ["foo", "bar"]}, "*", {"value": 11} ]
  },
  "@aRefExpr": {
    "=": "foo.bar",
    "ref": ["foo", "bar"]
  },
  "@aValueExpr": {
    "=": "11",
    "val": 11
  }
}
```

Note the different CSN representations for a plain value `"@anInteger": 11` and a value written
as expression `@aValueExpr: ( 11 )`, respectively.

#### Propagation

[Annotations are propagated](#annotation-propagation) in views/projections, in includes, or along type references.
Currently, paths are not rewritten in propagated annotations. In a projection, for example,
all elements used in an annotation expression must be projected without renaming.
Thus, for the time being we recommend to use the feature mainly in top-level projections.

Example:
<!-- cds-mode: upcoming, cds-compiler v4.5 -->
```cds
@Foo1 : (a)
@Foo2 : (a + b)

> @Foo3 : (a + c)

Reviewer: Maybe use names that are more "bookshop friendly"? 😄

Reviewer 2: If that is possible I'd be in favor of it :)

entity E {
  a : Integer;
  b : Integer;
  c : Integer;
}

entity P as projection on E {
  a,
  b as x
}
```

All three annotations are propagated to the projection `P`. The propagated `@Foo1` is still valid.
The propagated annotation `@Foo2` is invalid, as the referenced element `b` has been renamed to `x`.
The propagated annotation `@Foo3` is invalid, as element `c` of `E` has not been projected at all.
This results in a compiler error. To make it work, you would have to explicitly overwrite annotations
`@Foo2` and `@Foo3` at `P`.

> The compiler is going to take care of renamings and rewrite references in propagated annotations

Author: ```suggestion
The compiler is going to take care of renamed elements and rewrites references in propagated annotations
```

> The compiler is going to take care of renamings and rewrite references in propagated annotations

Reviewer: ```suggestion
The compiler is going to take care of renaming and rewriting references in propagated annotations
```

Author: That's not precisely what I want to say ... I'll try again ...

in a later release. The CSN representation of propagated annotation expressions may change even
if today no error is issued. Propagated annotation expressions that today are accepted may lead
to an error in the future when the implementation is improved.

#### CDS Annotations

Using an expression as annotation value only makes sense if the evaluator of the annotation is
prepared to deal with the new CSN representation.

> Currently the CAP runtimes only support expressions in the `where` property of the `@restrict` annotation.

Reviewer: ```suggestion
Currently, the CAP runtimes only support expressions in the `where` property of the `@restrict` annotation.
```

<!-- cds-mode: upcoming, cds-compiler v4.5 -->
```cds
entity Orders @(restrict: [
    { grant: 'READ', to: 'Auditor', where: (AuditBy = $user.id) }
  ]) {/*...*/}
```

More annotations are going to follow in upcoming releases.

> Of course you can use this feature also in your custom annotations, where you control the code that evaluates

Reviewer: ```suggestion
Of course, you can use this feature also in your custom annotations, where you control the code that evaluates
```

the annotations.

#### OData Annotations

The OData backend of CAP doesn't yet support expression valued annotations. This is planned for
a later release.
If you use the new expression syntax for OData annotations, the expression and contained references

> are not correctly translated, and the resulting EDMX will change once the OData support is available.

Reviewer: Well, in fact, you get a compilation error as of now 😄

Author: Not necessarily .

by CDS parsers or linkers. They're interpreted and evaluated only on consumption-specific modules.
For example, for SAP Fiori models, it's the _4odata_ and _2edm(x)_ processors.
:::

### Expressions as Annotation Values (beta) {#expressions-as-annotation-values}

::: warning
Expressions in annotation values are released as beta feature.
We provide an early preview of this functionality to allow you to experiment with it and provide feedback.
The behavior may change in later releases, in particular:
the behavior and the CSN representation of paths in propagated annotations will change,
and the behavior of expressions in OData annotations will change.
:::

In order to use an expression as an annotation value, it must be enclosed in parentheses:
<!-- cds-mode: upcoming, cds-compiler v4.5 -->
```cds
@anExpression: ( foo.bar * 11 )
```

Syntactically, the same expressions are supported as in a select item or in the where clause of a query,
except subqueries. The expression can of course also be a single reference or a simple value:
<!-- cds-mode: upcoming, cds-compiler v4.5 -->
```cds
@aRefExpr: ( foo.bar )

> @aValueExpr: ( 11 )

Reviewer 3: Maybe also give examples where the expression is a predicate

is checked by the compiler.

#### CSN Representation

In CSN, the expression is represented as a record with two properties:
* A string representation of the expression is stored in property `=`.
* A tokenized representation of the expression is stored in one of the properties
`xpr`, `ref`, `val`, `func`, etc. (like if the expression was written in a query).

```json
{
  "@anExpression": {
    "=": "foo.bar * 11",
    "xpr": [ {"ref": ["foo", "bar"]}, "*", {"value": 11} ]
  },
  "@aRefExpr": {
    "=": "foo.bar",
    "ref": ["foo", "bar"]
  },
  "@aValueExpr": {
    "=": "11",
    "val": 11
  }
}
```

Note the different CSN representations for a [plain value `"@anInteger": 11`](#annotation-values) and a value written
as expression `@aValueExpr: ( 11 )`, respectively.

#### Propagation

[Annotations are propagated](#annotation-propagation) in views/projections, in includes, or along type references.
Currently, paths are not rewritten in propagated annotations. In a projection, for example,
all elements used in an annotation expression must be projected without renaming.
Thus, for the time being we recommend to use the feature mainly in top-level projections.

> Example:

Author: @bugwelle @renejeglinsky Do you like that one better?

Reviewer 2: To be honest, I understand it now better. Thanks :)

Reviewer: Better, thanks! :)
